src/mlpack/methods/reinforcement_learning/replay/episodic_replay.hpp

```diff
@@ -0,0 +1,249 @@
+  * @param max_episode_len The maximum episode length possible.
+  * @param dimension The dimension of an encoded state.
+  */
+  EpisodicReplay(const size_t capacity,
+                 const size_t max_episode_len,
+                 const size_t dimension = StateType::dimension) :
+      capacity(capacity),
+      position(0),
+      max_episode_len(max_episode_len),
+      full(false)
+  {
+    states.resize(capacity);
+    next_states.resize(capacity);
+    rewards.resize(capacity);
+    actions.resize(capacity);
+    isTerminal.resize(capacity);
+  }
+
+  /**
+  * Store the given experience.
+  *
+  * @param state Given state.
+  * @param action Given action.
+  * @param reward Given reward.
+  * @param nextState Given next state.
+  * @param isEnd Whether next state is terminal state.
+  */
+  void Store(const StateType& state,
+             ActionType action,
+             double reward,
+             const StateType& nextState,
+             bool isEnd)
+  {
+    if (clear)
+    {
+      states[position].clear();
+      actions[position].clear();
+      rewards[position].clear();
+      next_states[position].clear();
+      isTerminal[position].clear();
+      clear = false;
+    }
+
+    states[position].push_back(state.Encode());
+    actions[position].push_back(action);
+    rewards[position].push_back(reward);
+    next_states[position].push_back(nextState.Encode());
+    isTerminal[position].push_back(isEnd);
+
+    if (isEnd || states[position].size() == max_episode_len)
+    {
+      position++;
+      clear = true;
+    }
+    if (position == capacity)
+    {
+      full = true;
+      position = 0;
+    }
+  }
+
+  /**
+  * Get the number of episodes in the memory.
+  *
+  * @return Actual used memory size
+  */
+  const size_t Size()
+  {
+    if (states[position].size() == 0)
+    {
+      return full ? capacity : position;
+    }
+    return full ? capacity : (position+1);
+  }
+
+  /**
+  * Get the most recently added episode.
+  *
+  * @param sampledStates Sampled encoded states.
+  * @param sampledActions Sampled actions.
+  * @param sampledRewards Sampled rewards.
+  * @param sampledNextStates Sampled encoded next states.
+  * @param isTerminal Indicate whether corresponding next state is terminal
+  *        state.
+  * @param random Whether episode is sampled random or most recent episode
+  *        is sampled
+  */
+  void Sample(arma::mat& episodeStates,
+                      arma::icolvec& episodeActions,
+                      arma::colvec& episodeRewards,
+                      arma::mat& episodeNextStates,
+                      arma::icolvec& isTerminal,
+                      bool random = false)
+  {
+    int episodeNum = 0;
+    if (random)
+    {
+      size_t upperBound = full ? capacity : position;
+      int lo = 0;
+      int high = upperBound;
+      if (!upperBound)
+      {
+        high = 1;
+      }
+      episodeNum = math::RandInt(lo, high);
```

Comment:
Could me make it
```suggestion
      episodeNum = math::RandInt(0, high);
```

And remove `lo` entirely?